splitgraph/cloud/tunnel_client.py

```diff
@@ -0,0 +1,90 @@
+import os
+import subprocess
+import sys
+from os import path
+from typing import IO, Any, Dict, Optional, cast
+
+from splitgraph.core.repository import Repository
+
+RATHOLE_CLIENT_CONFIG_FILENAME = "rathole-client.toml"
+
+RATHOLE_CLIENT_CONFIG_TEMPLATE = """
+[client]
+remote_addr = "{tunnel_server_management_address}"
+
+[client.transport]
+type = "tls"
+
+[client.transport.tls]
+{trusted_root_line}
+hostname = "{tls_hostname}"
+
+[client.services."{namespace}/{repository}"]
+local_addr = "{local_address}"
+# token is provisioner JWT token
+token = "{provisioning_token}"
+
+"""
+
+
+def get_rathole_client_config(
+    tunnel_server_management_address: str,
+    tls_hostname: str,
+    local_address: str,
+    provisioning_token: str,
+    namespace: str,
+    repository: str,
+    trusted_root: Optional[str],
+) -> str:
+    trusted_root_line = f'trusted_root = "{trusted_root}"' if trusted_root else ""
```

Comment:
Unfortunately rathole client will not start without a trusted root line provided in the config, so this will not work against production domains. This is based on my testing and also you can see the relevant [code here](https://github.com/rapiz1/rathole/blob/main/src/transport/tls.rs)). We could easily fork this and modify it to make it optional, which according to Rust docs will use the system native root by default. 

But it's probably not robust to assume a trust store exists at a specific location defined in an environment variable. Instead, we can extract the trusted root certificate from `data.splitgraph.com` (or whichever hostname we end up using here). We can do this with some similar commands to those in our [documentation for extracting certs for SSL mode in psql](https://www.splitgraph.com/docs/sql-client-reference/connecting/ssl-mode), but with an additional step to convert it to pkcs12. I'm looking into that now.

Reply:
In fact it would probably be easiest if the authentication server provided this (trusted root cert) to the client in its initial response.

Reply:
Update: There is no need to encode anything in `pkcs` format in the client config. A `pem` file is sufficient. In dev I used our `rootCA.pem` successfully, which _should_ be equivalent to the root certificate we extract from `www.splitgraph.test` using the commands in our documentation.

In production, the root certificate is from letsencrypt, so we can use the commands from our docs to extract a pem file (or can obtain it from other publicly available sources). Unfortunately rathole client requires a valid `trusted_root` even though Rust's default `TlsConnector` behavior is to use the system trust store.

That's fine though, it's more robust/secure to specify one anyway, and we need it in dev. But it would be nice to avoid the "extract the root certificate" step. For that reason I'm in favor of including the root certificate in the response from the authenticator (or provisioner).

----

(sidenote: We could open a PR to rathole to default to system trust store when `type = tls` but `trusted_root` does not exist. AFAICT this would be a simple change to the [`TlsConnectorBuilder` code](https://github.com/rapiz1/rathole/blob/main/src/transport/tls.rs#L34-L46) to create a TlsConnector even when trust store is omitted, which should [default to using the system trust store](https://docs.rs/native-tls/latest/native_tls/struct.TlsConnectorBuilder.html#method.disable_built_in_roots) instead of skipping creation and [causing a panic when its caller tries to unwrap it](https://github.com/rapiz1/rathole/blob/main/src/transport/tls.rs#L96). )

Reply:
One more update: even `rootCA.pem` is not required. Instead, the `fullchain.pem` certificate for the relevant domain is sufficient. I was able to use our development certificate for `data.splitgraph.com/fullchain.pem` successfully from my local machine (although I'm not sure if that's because `rootCA.pem` is already in my system trust store?)

For reference, it looks like the certificate extracted from the live site is not equivalent to `rootCA.pem`, but rather is equivalent to the `fullchain.pem` for that domain. So for consistency with prod it might make sense to use this `fullchain.pem` in dev:

```
# fingerprint of root certificate on disk at rootCA.pem
❯ openssl x509 -noout -modulus -in ~/splitgraph-cloud/certs/rootCA.pem | openssl md5                                                                                                                                 
(stdin)= 5167da4a173dfa4a99394b1c28bc1cb5

# fingerprint of certificate extracted from live request to data.splitgraph.test
❯ echo | openssl s_client -connect data.splitgraph.test:443 2>/dev/null | openssl x509 -noout -modulus | openssl md5                                                                                                 
(stdin)= 8b9badd1ab9536f11f5de49d0293f8e2

# fingerprint of "fullchain" certificate ("intermediate certificate?") on disk at fullchain.pem
❯ openssl x509 -noout -modulus -in ~/splitgraph-cloud/certs/data.splitgraph.com/fullchain.pem | openssl md5                                                                                                          
(stdin)= 8b9badd1ab9536f11f5de49d0293f8e2
```

(this is a bit over my head)